release-19.1: engine: Build with non-executable stacks

bdarnell opened the pull request:
Backport 1/1 commits from #37939.

/cc @cockroachdb/release

Fixes #37991

---

Three lines of inscrutable magic (and plenty of comments) to avoid
tripping a complex series of heuristics resulting in a security risk.

Fixes #37885

Release note (security improvement): Stack memory used by CockroachDB
is now marked as non-executable, improving security and compatibility
with SELinux.

bdarnell committed:
engine: Build with non-executable stacks

Three lines of inscrutable magic (and plenty of comments) to avoid
tripping a complex series of heuristics resulting in a security risk.

Fixes #37885

Release note (security improvement): Stack memory used by CockroachDB
is now marked as non-executable, improving security and compatibility
with SELinux.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/38011)
<!-- Reviewable:end -->

bdarnell commented:
Looks like this change breaks the windows build (in a way that isn't picked up by CI) so we can't merge it yet.

bdarnell committed:
engine: Rename the magic slice assembly file

Assembly files are supposed to be run through the preprocessor only
when they end in a captial .S. On the linux build, it appears to run
through the preprocessor anyway, but this is not the case for the
windows build (note that our windows binary is cross-compiled from
linux; this is not about case-insensitivity of the windows
filesystem). The windows build is broken when this file has a
lowercase .s.

Renames that consist only of case changes are poorly supported by git
and other tools, so also change the base name of the file. Some tools
support a .sx extension for a less-subtle version of this distinction,
but go does not.

Release note: None

bdarnell commented:
Added the commit from #38021 to fix the windows build.

bdarnell commented:
Ping
